### PR TITLE
Support Wasi as target

### PR DESCRIPTION
**Pull request by Mec-iS**
This starts the effort of supporting Wasi so to allow a base implementation for VM-based environments. It will probably require a rethinking for the features as the current `js` feature is meant for in-browser usage while we want also to support the Wasi binary interface that is still VM-based but not limited to browsers.

This target use `cargo-wasi` and `wasmtime` to run tests, see [doc](https://bytecodealliance.github.io/cargo-wasi/install.html).

cc: @morenol

### CODE REVIEW

**Comment by morenol** [comment]
I think that is misconfiguration. We should use `ubuntu` as os, to run the tests in a linux machine

**Comment by Mec-iS** [comment]
Thanks @morenol

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -56,3 +60,6 @@ jobs:
       - name: Tests in WASM
         if: matrix.platform.target == 'wasm32-unknown-unknown'
         run: wasm-pack test --node -- --all-features
+      - name: Tests in WASI
+        if: matrix.platform.target == 'wasm32‑wasi'
```
Comment (morenol) [inline .github/workflows/ci.yml:64]:
```suggestion
        if: matrix.platform.target == 'wasm32-wasi'
```

**Comment by Mec-iS** [comment]
Looks like that wasmtime is installed but needs to be exported in PATH.

Also we should add real time kernel for ROS or Raspberry Pi like in https://github.com/ros-realtime/linux-real-time-kernel-builder/blob/master/.github/workflows/rpi4-kernel-build.yml

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -42,6 +43,9 @@ jobs:
       - name: Install test runner for wasm
         if: matrix.platform.target == 'wasm32-unknown-unknown'
         run: curl https://rustwasm.github.io/wasm-pack/installer/init.sh -sSf | sh
+      - name: Install test runner for wasi
```
Comment (morenol) [inline .github/workflows/ci.yml:46]:
```suggestion
      - name: Install wasmtime
         if: matrix.platform.target == 'wasm32-wasi'
         run: curl https://wasmtime.dev/install.sh -sSf | bash
      - name: Install test runner for wasi
```

**Comment by morenol** [comment]
> Looks like wasmtime is installed but needs to be exported in PATH.
> 
> Also we should add real time kernel for ROS or Raspberry Pi like in https://github.com/ros-realtime/linux-real-time-kernel-builder/blob/master/.github/workflows/rpi4-kernel-build.yml

ohh right, I wrongly added another change

**Comment by Mec-iS** [comment]
ok now it runs but we got an error:
```
thread 'main' panicked at 'unknown instruction LocalTee(LocalTee { local: Id { idx: 26248 } })', crates/wasm-interpreter/src/lib.rs:363:18
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: failed to process wasm at `/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm`
Error: failed to process wasm at `/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm`
Caused by:
    failed to execute "/home/runner/.cache/cargo-wasi/0.1.26/wasm-bindgen/0.2.83/wasm-bindgen" "/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm" "--keep-debug" "--out-dir" "/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/.tmptbvsFn" "--out-name" "foo"
        status: exit status: 101
```
that is the same I got in my local machine (ubuntu 20.04)

**Comment by morenol** [comment]
> ok now it runs but we got an error:
> 
> ```
> thread 'main' panicked at 'unknown instruction LocalTee(LocalTee { local: Id { idx: 26248 } })', crates/wasm-interpreter/src/lib.rs:363:18
> note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
> error: failed to process wasm at `/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm`
> Error: failed to process wasm at `/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm`
> Caused by:
>     failed to execute "/home/runner/.cache/cargo-wasi/0.1.26/wasm-bindgen/0.2.83/wasm-bindgen" "/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/smartcore-24c4f627581dc4d2.rustc.wasm" "--keep-debug" "--out-dir" "/home/runner/work/smartcore/smartcore/target/wasm32-wasi/debug/deps/.tmptbvsFn" "--out-name" "foo"
>         status: exit status: 101
> ```
> 
> that is the same I got in my local machine (ubuntu 20.04)
> 
> With backtrace in my local:
> 
> ```
> thread 'main' panicked at 'unknown instruction LocalTee(LocalTee { local: Id { idx: 26251 } })', crates/wasm-interpreter/src/lib.rs:363:18
> stack backtrace:
>    0:     0x7f7b8ffbcf5d - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h32d04e59e827f4fb
>    1:     0x7f7b8fff356c - core::fmt::write::hc90753bb4142f112
>    2:     0x7f7b8ffba791 - std::io::Write::write_fmt::ha52e2782f4a3bd92
>    3:     0x7f7b8ffbe585 - std::panicking::default_hook::{{closure}}::hb0388c1054df5ebe
>    4:     0x7f7b8ffbe2a6 - std::panicking::default_hook::hb4a5679f0641e8af
>    5:     0x7f7b8ffbeb16 - std::panicking::rust_panic_with_hook::h55da91f28ae9c97c
>    6:     0x7f7b8ffbea07 - std::panicking::begin_panic_handler::{{closure}}::ha6d5d75e3abde1ad
>    7:     0x7f7b8ffbd414 - std::sys_common::backtrace::__rust_end_short_backtrace::hae84a849d394aed2
>    8:     0x7f7b8ffbe739 - rust_begin_unwind
>    9:     0x7f7b8fd14993 - core::panicking::panic_fmt::hf5f62850456bf94c
>   10:     0x7f7b8fe7db52 - wasm_bindgen_wasm_interpreter::Interpreter::call::h039daa71f050d7a2
>   11:     0x7f7b8fe7d8b1 - wasm_bindgen_wasm_interpreter::Interpreter::call::h039daa71f050d7a2
>   12:     0x7f7b8fe7d8b1 - wasm_bindgen_wasm_interpreter::Interpreter::call::h039daa71f050d7a2
>   13:     0x7f7b8fe7d8b1 - wasm_bindgen_wasm_interpreter::Interpreter::call::h039daa71f050d7a2
>   14:     0x7f7b8fe7cbe7 - wasm_bindgen_wasm_interpreter::Interpreter::interpret_descriptor::h89cd2766533f4dad
>   15:     0x7f7b8fd2e1d9 - wasm_bindgen_cli_support::Bindgen::generate_output::hb7c7f525f0bc2bd3
>   16:     0x7f7b8fd1abae - wasm_bindgen_cli_support::Bindgen::generate::hb4f43a5f3a936189
>   17:     0x7f7b8fd1808e - wasm_bindgen::main::h051811b559acfdbe
>   18:     0x7f7b8fd15c53 - std::sys_common::backtrace::__rust_begin_short_backtrace::h8707b38fb4f51a3d
>   19:     0x7f7b8fd15eb9 - std::rt::lang_start::{{closure}}::h59eb0de2404d2f2a
>   20:     0x7f7b8ffb5642 - std::rt::lang_start_internal::h37683b205c9af4e4
>   21:     0x7f7b8fd185a2 - main
> error: failed to process wasm at `/home/lorenzo/code/rust/smartcore/target/wasm32-wasi/debug/deps/smartcore-a1d4ac85eacbd7a4.rustc.wasm`
> ```
> 
> looks like doesn't like fs writing like any other VM.
> 
> It can be `use std::fs;`, we may need to exclude `readears` for Wasi.

it looks like it is due to `wasm-bindgen` dependency added in tests. I added commits to fix that

**Comment by Mec-iS** [comment]
great!

**Review by morenol — APPROVED** [review]

**Pull request merged**